### PR TITLE
Generic ECC Chips

### PR DESCRIPTION
**Pull request by 4rmanc**

### CODE REVIEW

File: circuit/src/ecc/generic/native.rs [hunk 1]
```diff
@@ -274,6 +274,81 @@ where
...
+{
+	fn without_witnesses() -> Self {
+		Self {
+			_ec: PhantomData,
+			x: UnassignedInteger::without_witnesses(),
+			y: UnassignedInteger::without_witnesses(),
+		}
+	}
+}
+
```
Comment (lazovicff) [inline circuit/src/ecc/generic/native.rs:351]: This should be in mod.rs

File: circuit/src/ecc/generic/mod.rs [hunk 2]
```diff
@@ -1,2 +1,1798 @@
...
+
+/// Configuration elements for the circuit are defined here.
+#[derive(Debug, Clone)]
+pub struct EccAddConfig {
+	/// Constructs selectors from different circuits.
+	integer_reduce_selector: Selector,
+	integer_sub_selector: Selector,
+	integer_mul_selector: Selector,
+	integer_div_selector: Selector,
+}
```
Comment (lazovicff) [inline circuit/src/ecc/generic/mod.rs:67]: We should move this and every other config inside ecc/mod.rs file and use it for both same_curve and generic

File: circuit/src/ecc/generic/native.rs (hunk 1 above)
Comment (lazovicff) [reply circuit/src/ecc/generic/native.rs:351]: I mean generic/mod.rs

**Comment by 4rmanc** [comment]
@lazovicff Circuit didn't returned proper results compare to native version. This was because native version returns directly bits `(length = 256)` when circuit returns limb bits `(length = 272)`. Instead changing `bits2integer` I [changed native version's bit calculation](https://github.com/eigen-trust/protocol/blob/generic_ecc_chips/circuit/src/ecc/generic/native.rs#L170-L176). What do you think?

File: circuit/src/ecc/generic/native.rs (hunk 1 above)
Comment (4rmanc) [reply circuit/src/ecc/generic/native.rs:355]: I moved all the `Unassigned...` types to `mod.rs` files

File: circuit/src/ecc/generic/mod.rs [hunk 3]
```diff
@@ -1,2 +1,1754 @@
...
+pub struct UnassignedEcPoint<
+	C: CurveAffine,
+	N: FieldExt,
+	const NUM_LIMBS: usize,
+	const NUM_BITS: usize,
+	P,
+	EC,
+> where
+	P: RnsParams<C::Base, N, NUM_LIMBS, NUM_BITS> + RnsParams<C::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	EC: EccParams<C>,
```
Comment (lazovicff) [inline circuit/src/ecc/generic/mod.rs:42]: Looks like `EC: EccParams<C>,` is not needed for UnassignedEcPoint?

File: circuit/src/ecc/generic/mod.rs [hunk 4]
```diff
@@ -1,2 +1,1754 @@
...
+			_ec: PhantomData,
+		}
+	}
+}
+
+impl<C: CurveAffine, N: FieldExt, const NUM_LIMBS: usize, const NUM_BITS: usize, P, EC>
+	UnassignedValue for UnassignedEcPoint<C, N, NUM_LIMBS, NUM_BITS, P, EC>
+where
+	P: RnsParams<C::Base, N, NUM_LIMBS, NUM_BITS> + RnsParams<C::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	EC: EccParams<C>,
```
Comment (lazovicff) [inline circuit/src/ecc/generic/mod.rs:93]: Looks like `EC` is not needed in `UnassignedEcPoint`?

File: circuit/src/ecc/generic/mod.rs [hunk 5]
```diff
@@ -1,2 +1,1754 @@
...
+/// Structure for the AssignedPoint.
+#[derive(Clone, Debug)]
+pub struct AssignedPoint<
+	C: CurveAffine,
+	N: FieldExt,
+	const NUM_LIMBS: usize,
+	const NUM_BITS: usize,
+	P,
+> where
+	P: RnsParams<C::Base, N, NUM_LIMBS, NUM_BITS> + RnsParams<C::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
```
Comment (lazovicff) [inline circuit/src/ecc/generic/mod.rs:115]: No need for ` + RnsParams<C::ScalarExt, N, NUM_LIMBS, NUM_BITS>`

File: circuit/src/ecc/generic/mod.rs [hunk 6]
```diff
@@ -1,2 +1,1754 @@
...
+	<C as CurveAffine>::Base: FieldExt,
+	<C as CurveAffine>::ScalarExt: FieldExt,
```
Comment (lazovicff) [inline circuit/src/ecc/generic/mod.rs:117]: No need for `<C as CurveAffine>::ScalarExt: FieldExt,`

File: circuit/src/ecc/generic/mod.rs [hunk 7]
```diff
@@ -1,2 +1,1754 @@
...
+pub struct EccAddChipset<
+	C: CurveAffine,
+	N: FieldExt,
+	const NUM_LIMBS: usize,
+	const NUM_BITS: usize,
+	P,
+> where
+	P: RnsParams<C::Base, N, NUM_LIMBS, NUM_BITS> + RnsParams<C::ScalarExt, N, NUM_LIMBS, NUM_BITS>,
+	<C as CurveAffine>::Base: FieldExt,
+	<C as CurveAffine>::ScalarExt: FieldExt,
```
Comment (lazovicff) [inline circuit/src/ecc/generic/mod.rs:151]: Same for chipsets.
No need for ` + RnsParams<C::ScalarExt, N, NUM_LIMBS, NUM_BITS>` and `<C as CurveAffine>::ScalarExt: FieldExt,` except for scalar multiplication chipsets

File: circuit/src/ecc/generic/native.rs [hunk 8]
```diff
@@ -191,7 +199,7 @@ where
 	pub fn multi_mul_scalar(
 		points: &[Self], scalars: &[Integer<C::ScalarExt, N, NUM_LIMBS, NUM_BITS, P>],
 	) -> Vec<Self> {
-		let sliding_window_size = EC::window_size();
+		let sliding_window_size = 2;
```
Comment (lazovicff) [inline circuit/src/ecc/generic/native.rs:202]: Is this a leftover?

**Comment by lazovicff** [comment]
Thanks for updating unassigned values

File: circuit/src/ecc/generic/native.rs (hunk 8 above)
Comment (4rmanc) [reply circuit/src/ecc/generic/native.rs:202]: Yes, fixed it.

File: circuit/src/ecc/generic/mod.rs (hunk 3 above)
Comment (4rmanc) [reply circuit/src/ecc/generic/mod.rs:42]: Yes, fixed.

**Review by lazovicff — APPROVED** [review]

**Pull request merged**